### PR TITLE
[#5] FEAT : RNN 알고리즘 및 라이트닝 인프라 Baseline 작성

### LINKED ISSUE #5: [FEAT] RNN(LSTM) 라이트닝 구현
이슈 내용 요약: 모델 구축하고 트레이닝 코드 구현
- input : 과거 데이터 512개(임의의 값 : 특정 시점 바로 직전)
- output : 미래 데이터 100개(임의의 값 : 특정 시점 바로 직후)의 평균값
- purpose : 특정시점을 기준으로 살것인지 팔것인지를 판단하는데 도움을 주기 위해서 특정시점을 기준으로 나눔
진행 예정 사항
- [x] 예정 사항 1. Lightning 틀 만들기
- [x] 예정 사항 2. RNN 모델(LSTM) 만들기
- [x] 예정 사항 2-1. Dataset 모듈 만들기(sliding window)
- [x] 예정 사항 3. Dataloader 만들기

### PR DESCRIPTION
**Pull request by eno3940**
## 세부 내용

- 딥러닝 라이트닝 인프라 구축
- 특정 시점에 주식을 판매할지, 구매할지를 도와주는 알고리즘 : 특정 시점 이전 가격 데이터를 학습시켜 가까운 미래에서의 평균 주식값을 예측
- Dataset 모듈 작성 (가격 데이터를 slidingwindow방식으로 input과 label값 분리)
- Datamodule 모듈 작성
- model 모듈 작성
- 훈련에 필요한 기본적인 틀만 작성. 추후 평가지표 설정, 테스트, 디벨롭 필요

## 이슈 번호
- Resolved: #5

### CODE REVIEW

File: .DS_Store
Comment (dkqp) [inline .DS_Store:1]: .DS_Store 파일은 삭제하고, gitignore 파일에 .DS_Store 를 추가해서 해당 파일이 추적되지 않도록 하면 좋을것 같아요.

File: DL_algorithms/StockDataset.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        self.prediction_length = prediction_length
+        self.data = pd.read_csv(data_dir)
+
+        self.processed_data, self.labels = self._preprocess()
+
+    def _preprocess(self):
+        processed_data = []
+        labels = []
+        
+        for start in range(0, len(self.data) - self.window_size - self.prediction_length + 1, self.step_size):
```
Comment (dkqp) [inline DL_algorithms/StockDataset.py]: + 1 이 들어가면 end + prediction_length 가 len(self.data) 를 초과하는 엣지 케이스가 생길것 같은데 다시 검토해봐주세요. 만약 그렇다고 해도 에러는 딱히 안나겠지만 의도한 부분은 아닐테니 빼는게 좋을것 같네요 ㅎㅎ

File: DL_algorithms/StockDatamodule.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from lightning.pytorch.utilities.types import EVAL_DATALOADERS, TRAIN_DATALOADERS, STEP_OUTPUT
+from torch.utils.data import DataLoader, random_split
+
+class StockDataModule(pl.LightningDataModule):
+    def __init__(self, batch_size: int = 32, data_dir:str = '../data', window_size:int = 512, step_size:int=20, prediction_length:int = 100) -> None:
```
Comment (dkqp) [inline DL_algorithms/StockDatamodule.py]: data_dir 에 따로 값이 입력되지 않아 '../data' 라는 default 값이 입력되는 경우에는 결국 에러가 발생하지 않을까요?
값을 항상 넣어줘야 하는 변수로 두고, batch_size 의 앞으로 빼는게 좋겠네요 ㅎㅎ
그리고 변수명도 data_path 정도가 더 적당하지 않을까 싶구요.

File: DL_algorithms/StockDatamodule.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            train_dataset = SlidingWindowDataset(
+                data_dir=self.data_dir,
+                window_size= self.window_size,
+                step_size= self.step_size,
+                prediction_length= self.prediction_length
+            )
+
+
+            # split train val test dataset
+            self.train_dataset, self.val_dataset, self.test_dataset = random_split(
```
Comment (dkqp) [inline DL_algorithms/StockDatamodule.py]: 이전에 논의한 내용 중 하나이긴 한데, 20 step 간격으로 데이터를 만들었기 때문에 이를 random 하게 split 하게 되면 validate 또는 test 하려는 데이터와 유사한 데이터가 이미 train 단계에서 학습되었을 가능성이 높을것 같아요.
그래서 그보다는 과거 데이터로 train 하고 최근 데이터로 validate 또는 test 를 하는게 좋지 않을까 생각해요. 순서에 따라 특정 비율로 dataset 을 나눌 수 있는 방법을 찾아서 시도해봐주세요.

File: DL_algorithms/Stockmodel.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        y_hat = self(x)
+        loss = F.mse_loss(y_hat, y)
+
+        metrics = {
+            'test_loss': loss,
+        }
+        self.log_dict(metrics, prog_bar=True)
+
+    def predict_step(self, batch, batch_idx):
+        return torch.argmax(self.model(batch), dim=-1)
```
Comment (dkqp) [inline DL_algorithms/Stockmodel.py]: argmax 는 분류할때 사용하는데, 우리는 regression 문제에 맞춰서 예측값을 내주어야 할 것 같아요 ㅎ

File: DL_algorithms/StockDataset.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            end = start + self.window_size
+            window_data = self.data[start:end]['o']
+            label_data = self.data[end:end + self.prediction_length]['o'].mean()
```
Comment (dkqp) [inline DL_algorithms/StockDataset.py:21]: 현재는 input 과 label 데이터 각각이 원래값 그대로 들어가도록 되어있는데요,
사실 input 값이 동일한 변화를 보이는데 다만 전체적으로 가격이 2배로 뛰었다면, label 역시 2배로 뛰었다는 결과를 내는 것이 자연스러울 것 같아요.
특정값 (예를 들어 '현재 시점 가격') 을 기준으로 상대적인 가격으로 데이터를 표현하는 것이 학습에 유리하지 않을까요?

**Review by dkqp — CHANGES_REQUESTED** [review]
코드 깔끔한게 보기 좋네요 ㅎㅎ 고생했어요.
제가 의견 단 부분 검토해보시고 댓글 달아주시거나, 필요한 부분 수정해서 commit & push 부탁해요~!

**Review by dkqp — CHANGES_REQUESTED** [review]
수고하셨습니다~! 이제 학습이 잘 되는지 확인해보면 될것 같네요.
다만 현재 csv 파일과 라이트닝 로그 파일들이 모두 올라와 있습니다. 코드와 주요 파일 외에는 로컬에서만 관리하도록 .gitignore 를 수정하고 나서 머지하시죠.

**Review by dkqp — APPROVED** [review]
